python/ray/data/_internal/block_batching/iter_batches.py

```diff
@@ -0,0 +1,311 @@
+            slice_index = slice_indices[i]
+            output.add_block(
+                accessor.slice(
+                    slice_index[0],
+                    slice_index[1]
+                    if slice_index[1] is not None
+                    else accessor.num_rows(),
+                    copy=False,
+                )
+            )
+
+        batch = output.build()
+        assert len(batch) == logical_batch.num_rows, (
+            len(batch),
+            logical_batch.num_rows,
+        )
+        if ensure_copy:
+            # Need to ensure that the batch is a fresh copy.
+            batch = BlockAccessor.for_block(batch)
+            batch = batch.slice(0, batch.num_rows(), copy=True)
+
+        yield Batch(logical_batch.batch_idx, batch, logical_batch)
+
+
+def format_batches(
+    block_iter: Iterator[Batch],
+    batch_format: Optional[str],
+) -> Iterator[Batch]:
+    """Given an iterator of blocks, returns an iterator of formatted batches.
+
+    Args:
+        block_iter: An iterator over blocks.
+        batch_format: The batch format to use.
+        stats: An optional stats object to record formatting times.
+
+    Returns:
+        An iterator over batch index and the formatted batch.
+    """
+    for batch in block_iter:
+        formatted_batch = BlockAccessor.for_block(batch.data).to_batch_format(
+            batch_format
+        )
+        batch.data = formatted_batch
+        yield batch
+
+
+def collate(
+    batch_iter: Iterator[Batch],
+    collate_fn: Optional[Callable[[DataBatch], Any]],
+) -> Iterator[Batch]:
+    """Returns an iterator with the provided collate_fn applied to items of the batch
+    iterator.
+
+    Args:
+        batch_iter: An iterator over formatted batches.
+        stats: An optional stats object to record collation time.
+    """
+    for batch in batch_iter:
+        batch.data = collate_fn(batch.data)
+        yield batch
+
+
+def trace_deallocation_for_batch(
+    batch_iter: Iterator[Batch], eager_free: bool
+) -> Iterator[Batch]:
+    """Trace deallocation of the underlying block references for each batch.
+
+    Args:
+        batch_iter: An iterator over batches.
+        eager_free: Whether to eagerly free the object reference from the object store.
+    """
+    for batch in batch_iter:
+        block_refs = batch.logical_batch.block_refs
+        for block_ref in block_refs:
+            trace_deallocation(block_ref, loc="iter_batches", free=eager_free)
+        yield batch
+
+
+def restore_from_original_order(batch_iter: Iterator[Batch]) -> Iterator[Batch]:
+    """Restores the original order of the provided `batch_iter`
+
+    This function will yield items from `base_iterator` in the correct order based on
+    each batch's batch_idx. All indexes are expected to be unique.
+
+    `batch_iter` is expected to not have any missing indexes. All indexes from 0 to len
+    (base_iterator) must be present.
+    """
+    next_index_required = 0
+    buffer: List[Batch] = []
```

Comment:
I think using a Dict[index, batch] here would be simpler.

Reply:
good point, updated